### PR TITLE
drivers: clock_control: nrf: Add API for synchronous request

### PR DESCRIPTION
**Pull request by nordic-krch**
Add API for synchronous request for clock attributes.

### CODE REVIEW

**Review by gmarull — DISMISSED** [review]

File: drivers/clock_control/clock_control_nrf2_common.c [hunk 1]
```diff
@@ -159,3 +167,37 @@ int api_nosys_on_off(const struct device *dev, clock_control_subsys_t sys)
...
+	k_sem_init(&sem, 0, 1);
+	data.sem = &sem;
+	sys_notify_init_callback(&data.cli.notify, sync_cb);
+
+	err = nrf_clock_control_request(dev, spec, &data.cli);
+	if (err < 0) {
+		return err;
+	}
+
+	err = k_sem_take(&sem, K_MSEC(CONFIG_CLOCK_CONTROL_NRF2_NRFS_CLOCK_TIMEOUT_MS));
```
Comment (anangl) [inline drivers/clock_control/clock_control_nrf2_common.c:197]: I don't think that's reasonable. Not all clock controllers use nrfs. This timeout should be rather specified with the parameter of the API function.

File: drivers/clock_control/clock_control_nrf2_common.c [hunk 2]
```diff
@@ -159,3 +167,37 @@ int api_nosys_on_off(const struct device *dev, clock_control_subsys_t sys)
...
+	k_sem_give(data->sem);
+}
+
+int nrf_clock_control_request_sync(const struct device *dev, const struct nrf_clock_spec *spec)
+{
+	struct clock_control_nrf_sync_req_data data;
+	struct k_sem sem;
+	int err;
+
+	__ASSERT_NO_MSG(!k_is_in_isr());
```
Comment (anangl) [inline drivers/clock_control/clock_control_nrf2_common.c:186]: How about returning `-EWOULDBLOCK` here?

File: drivers/clock_control/clock_control_nrf2_common.c [hunk 3]
```diff
@@ -159,3 +167,37 @@ int api_nosys_on_off(const struct device *dev, clock_control_subsys_t sys)
...
+	struct clock_control_nrf_sync_req_data  *data =
+		CONTAINER_OF(cli, struct clock_control_nrf_sync_req_data, cli);
+
+	data->res = res;
+	k_sem_give(data->sem);
+}
+
+int nrf_clock_control_request_sync(const struct device *dev, const struct nrf_clock_spec *spec)
+{
+	struct clock_control_nrf_sync_req_data data;
```
Comment (anangl) [inline drivers/clock_control/clock_control_nrf2_common.c:182]:
```suggestion
	struct clock_control_nrf_sync_req_data req;
```

File: drivers/clock_control/clock_control_nrf2_common.c [hunk 4]
```diff
@@ -24,6 +25,13 @@ LOG_MODULE_REGISTER(clock_control_nrf2, CONFIG_CLOCK_CONTROL_LOG_LEVEL);
  */
 STRUCT_CLOCK_CONFIG(generic, ONOFF_CNT_MAX);
 
+/* Structure used for synchronous clock request. */
+struct clock_control_nrf_sync_req_data {
```
Comment (anangl) [inline drivers/clock_control/clock_control_nrf2_common.c:29]: Why not name it just `sync_req`? It's a local structure.

File: drivers/clock_control/clock_control_nrf2_common.c [hunk 5]
```diff
@@ -159,3 +167,37 @@ int api_nosys_on_off(const struct device *dev, clock_control_subsys_t sys)
...
+int nrf_clock_control_request_sync(const struct device *dev, const struct nrf_clock_spec *spec)
+{
+	struct clock_control_nrf_sync_req_data data;
+	struct k_sem sem;
+	int err;
+
+	__ASSERT_NO_MSG(!k_is_in_isr());
+
+	k_sem_init(&sem, 0, 1);
+	data.sem = &sem;
```
Comment (anangl) [inline drivers/clock_control/clock_control_nrf2_common.c:189]: Why not make this semaphore a member of this structure? And initialize it statically?

**Review by anangl — CHANGES_REQUESTED** [review]

**Comment by nordic-krch** [comment]
@bjarki-andreasen @anangl comments applied.

**Review by bjarki-andreasen — DISMISSED** [review]
Looks good

**Review by anangl — DISMISSED** [review]

File: include/zephyr/drivers/clock_control/nrf_clock_control.h [hunk 6]
```diff
@@ -235,6 +235,23 @@ int nrf_clock_control_request(const struct device *dev,
 	return api->request(dev, spec, cli);
 }
 
+/**
+ * @brief Synchronously request a reservation to use a given clock with specified attributes.
```
[the remark this answers is not in the document]
Comment (nordic-krch) [reply include/zephyr/drivers/clock_control/nrf_clock_control.h:239]: Added a note that clarifies that.

**Comment by nordic-krch** [comment]
@bjarki-andreasen @gmarull can you take another look? Only note added in the function description.

**Review by bjarki-andreasen — APPROVED** [review]

**Review by masz-nordic — APPROVED** [review]

**Review by anangl — APPROVED** [review]

**Review by kl-cruz — APPROVED** [review]

**Pull request merged**